src/main/java/org/voltdb/exportclient/KinesisFirehoseExportClient.java

```diff
@@ -0,0 +1,264 @@
+            } catch(IOException e) {
+                rateLimitedLogError(m_logger, "Failed to build record: %s", Throwables.getStackTraceAsString(e));
+                throw new RestartBlockException(true);
+            } finally {
+                try { csv.close(); } catch (IOException e) {}
+            }
+
+            m_records.add(record);
+            return true;
+        }
+
+        @Override
+        public void sourceNoLongerAdvertised(AdvertisedDataSource source)
+        {
+            m_es.shutdown();
+            try {
+                m_es.awaitTermination(365, TimeUnit.DAYS);
+            } catch (InterruptedException e) {
+                Throwables.propagate(e);
+            }
+        }
+
+        @Override
+        public void onBlockStart() throws RestartBlockException
+        {
+            m_records = new ArrayList<Record>();
+        }
+
+        @Override
+        public void onBlockCompletion() throws RestartBlockException
+        {
+            try {
+                int recordsSize = m_records.size();
+                List<Record> recordsList;
+                int sleepTime = 0;
+                while (recordsSize > 0) {
+                    if (sleepTime > 0)
+                        Thread.sleep(sleepTime);
+                    // PutRecordBatchRequest can not contain more than 500 records
+                    if (recordsSize > 500) {
+                        recordsList = m_records.subList(0, 499);
+                        m_records = new ArrayList<Record>(m_records.subList(500, recordsSize-1));
+                        recordsSize = m_records.size();
+                    } else {
+                        recordsList = new ArrayList<Record>(m_records);
+                        m_records.clear();
+                        recordsSize = 0;
+                    }
+                    PutRecordBatchRequest batchRequest = new PutRecordBatchRequest().
+                            withDeliveryStreamName(m_streamName).
+                            withRecords(recordsList);
+                    PutRecordBatchResult res = m_firehoseClient.putRecordBatch(batchRequest);
+                    if (res.getFailedPutCount() > 0) {
+                        for (PutRecordBatchResponseEntry entry : res.getRequestResponses()) {
+                            if (entry.getErrorMessage() != null && !entry.getErrorMessage().contains("Slow down.")) {
+                                rateLimitedLogError(m_logger, "Record failed with response: %s", entry.getErrorMessage());
+                                throw new RestartBlockException(true);
+                            }
+                        }
+                        m_records.addAll(0, recordsList);
+                        recordsSize = m_records.size();
+                        sleepTime = sleepTime == 0 ? 1000 : sleepTime*2;
+                    } else
+                        sleepTime = sleepTime == 0 ? 0 : sleepTime-10;
```

Comment:
I'm not sure, 10 seemed reasonable to me since this should only happen under extremely heavy workload. For example in my testing I used voter with the votes table exporting. I was seeing that each block size was around 30,000 so for this there would be more than 60 groups of 500. Since there are so many of these and we want to avoid doubling the sleep time I think it should reduce the sleep time slowly.